apps/platform/src/journey/JourneyService.ts

```diff
@@ -1,93 +1,296 @@
+                if (userStep.type === 'completed') {
+                    step = await this.nextOrEnd(step)
+                    continue
+                }
+
+                // exit journey completely if a catastrophic error
+                // has occurred to avoid unpredictable behavior
+                if (userStep.type === 'error') {
+                    await this.end()
+                }
 
-        let nextStep: JourneyStep | undefined | null = await this.nextStep(user)
-        while (nextStep) {
-            if (processed.includes(nextStep.id)) {
-                // Avoid infinite loop in single run
+                // stop processing for now
                 break
             }
-            processed.push(nextStep.id)
-            nextStep = this.parse(nextStep)
-
-            // If completed, jump to next otherwise validate condition
-            if (await nextStep.hasCompleted(user)) {
-                nextStep = await nextStep.next(user)
-            } else if (await nextStep.condition(user, event)) {
-                const proceed = await nextStep.complete(user, event)
-                if (proceed) {
-                    nextStep = await nextStep.next(user)
-                } else {
-                    nextStep = null
-                }
-            } else {
-                nextStep = null
+
+            // delegate to step type
+            await step.process(this, userStep)
+
+            // persist and update the user step
+            userStep.parseJson(
+                userStep.id
+                    ? await JourneyUserStep.updateAndFetch(userStep.id, userStep)
+                    : await JourneyUserStep.insertAndFetch(userStep),
+            )
+        }
+
+        if (this._jobs.length) {
+            await App.main.queue.enqueueBatch(this._jobs.map(j => typeof j === 'function' ? j(this) : j))
+        }
+    }
+
+    private async nextOrEnd(step: JourneyStep) {
+        const stepId = await step.next(this)
+        if (stepId) {
+            const step = this.steps.find(s => s.id === stepId)
+            if (step) {
+                return step
             }
         }
+        await this.end()
+    }
+
+    private async end() {
+        await JourneyUserStep.update(q => q.where('id', this.entrance.id), {
+            ended_at: new Date(),
+        })
     }
 
-    parse(step: JourneyStep): JourneyStep {
-        return journeyStepTypes[step.type]?.fromJson(step)
+    public childrenOf(stepId: number) {
+        return this.children.filter(sc => sc.step_id === stepId)
     }
 
-    async nextStep(user: User): Promise<JourneyStep | undefined> {
+    public job(job: JobOrJobFunc) {
+        this._jobs.push(job)
+    }
 
-        // Get the ID of last step the user has started previously
-        const lastUserStep = await lastJourneyStep(user.id, this.journeyId)
+    public async events() {
+        // lazy load, only grab the specific event types that we need for gates in this journey.
+        if (!this._events) {
```

Comment:
Random stylistic suggestion, instead of going so deep there are lots of times like here we could just return instead. So do `if (this._events) return this.events` and then you don't go as nested and it's easier to read.